File: frontend/src/api/modules/screener.ts
import request from '../request'
import type { ScreenerCondition, AIParseResult, AIPickResult, AIAnalyzeResult, AIDailyRecommendation } from '@/types/screener'

const AI_TIMEOUT = 30000

const API_BASE = import.meta.env.VITE_API_BASE_URL || '/api'

export const screenerApi = {
  screen: (conditions: ScreenerCondition[], limit = 50) =>
    request.post('/screener', { conditions, limit }),
  getTemplates: () =>
    request.get('/screener/templates'),
  getIndustries: () =>
    request.get('/screener/industries'),
  aiParse: (query: string) =>
    request.post('/screener/ai-parse', { query }, { timeout: AI_TIMEOUT }) as Promise<AIParseResult>,
  aiPick: (query: string) =>
    request.post('/screener/ai-pick', { query }, { timeout: AI_TIMEOUT }) as Promise<AIPickResult>,
  aiAnalyze: (tsCodes: string[]) =>
    request.post('/screener/ai-analyze', { ts_codes: tsCodes }, { timeout: AI_TIMEOUT }) as Promise<AIAnalyzeResult>,
  aiDaily: () =>
    request.get('/screener/ai-daily', { timeout: AI_TIMEOUT }) as Promise<AIDailyRecommendation>,
  aiChat: (query: string, history: { role: string; content: string }[]) =>
    request.post('/screener/ai-chat', { query, history }, { timeout: AI_TIMEOUT }) as Promise<{
      text: string; conditions: any[]; industries: string[]; stocks: any[]; stock_count: number
    }>,

  /**
   * SSE 流式 AI 对话
   * @returns AbortController 用于中断连接
   */
  aiChatStream: (
    query: string,
    history: { role: string; content: string }[],
    callbacks: {
      onChunk: (text: string) => void
      onDone: (result: {
        text: string; conditions: any[]; industries: string[]
        stocks: any[]; stock_count: number
      }) => void
      onError: (message: string) => void
    },
  ): AbortController => {
    const controller = new AbortController()

    fetch(`${API_BASE}/screener/ai-chat/stream`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${localStorage.getItem('token') || ''}`,
      },
      body: JSON.stringify({ query, history }),
      signal: controller.signal,
    }).then(async (response) => {
      if (!response.ok) {
        callbacks.onError(`HTTP ${response.status}`)
        return
      }
      const reader = response.body?.getReader()
      if (!reader) {
        callbacks.onError('浏览器不支持流式读取')
        return
      }
      const decoder = new TextDecoder()
      let buffer = ''

      while (true) {
        const { done, value } = await reader.read()
        if (done) break
        buffer += decoder.decode(value, { stream: true })
        const lines = buffer.split('\n')
        buffer = lines.pop() || ''
        for (const line of lines) {
          if (line.startsWith('data: ')) {
            try {
              const event = JSON.parse(line.slice(6))
              if (event.type === 'chunk') {
                callbacks.onChunk(event.text)
              } else if (event.type === 'done') {
                callbacks.onDone(event)
              } else if (event.type === 'error') {
                callbacks.onError(event.text)
              }
            } catch {
              // 忽略解析错误
            }
          }
        }
      }
    }).catch((err) => {
      if (err.name !== 'AbortError') {
        callbacks.onError(err.message || '网络错误')
      }
    })

    return controller
  },
}

File: backend/app/api/routers/screener.py
from fastapi import APIRouter, Depends
from app.models.response import BaseResponse
from app.models.screener import ScreenerRequest, AIParseRequest, AIPickRequest, AIAnalyzeRequest, AIChatRequest
from app.services.screener_service import ScreenerService
from app.services.screener_ai_service import ScreenerAIService
from app.api.deps import get_db
from motor.motor_asyncio import AsyncIOMotorDatabase

router = APIRouter(prefix="/api/screener", tags=["筛选"])


async def get_screener_service(db: AsyncIOMotorDatabase = Depends(get_db)) -> ScreenerService:
    return ScreenerService(db)


async def get_ai_service(db: AsyncIOMotorDatabase = Depends(get_db)) -> ScreenerAIService:
    return ScreenerAIService(db)


@router.post("")
async def screen(
    req: ScreenerRequest,
    service: ScreenerService = Depends(get_screener_service),
):
    data = await service.screen(req.conditions, req.limit)
    return BaseResponse(data=data)


@router.get("/templates")
async def get_templates(
    service: ScreenerService = Depends(get_screener_service),
):
    data = await service.get_templates()
    return BaseResponse(data=data)


@router.get("/industries")
async def get_industries(
    service: ScreenerService = Depends(get_screener_service),
):
    data = await service.get_industries()
    return BaseResponse(data=data)


@router.post("/ai-parse")
async def ai_parse(
    req: AIParseRequest,
    ai_service: ScreenerAIService = Depends(get_ai_service),
):
    data = await ai_service.parse_natural_language(req.query)
    return BaseResponse(data=data)


@router.post("/ai-pick")
async def ai_pick(
    req: AIPickRequest,
    ai_service: ScreenerAIService = Depends(get_ai_service),
):
    data = await ai_service.ai_pick(req.query)
    return BaseResponse(data=data)


@router.post("/ai-analyze")
async def ai_analyze(
    req: AIAnalyzeRequest,
    ai_service: ScreenerAIService = Depends(get_ai_service),
):
    data = await ai_service.ai_analyze(req.ts_codes)
    return BaseResponse(data=data)


@router.get("/ai-daily")
async def ai_daily(
    ai_service: ScreenerAIService = Depends(get_ai_service),
):
    data = await ai_service.get_daily_recommendation()
    return BaseResponse(data=data)


@router.post("/ai-chat")
async def ai_chat(
    req: AIChatRequest,
    ai_service: ScreenerAIService = Depends(get_ai_service),
):
    data = await ai_service.ai_chat(req.query, req.history)
    return BaseResponse(data=data)
